### ISSUE TITLE
Searching with ":vimgrep" in files with certain filenames causes an E944

### ISSUE THREAD

**Issue by lkintact** [vim/vim#6919]
**To Reproduce**
1. Create a file with the name `[id-01] file.txt` and contents `test`.
2. Run `gvim --clean --cmd "language messages English" --cmd "set encoding=utf-8" "[id-01] file.txt"`.
3. Execute `:vimgrep /test/ %`.
4. gVim finds the string, but emits an `E944: Reverse range in character class` error.

**Expected behavior**
The error should not be emitted.

**Screenshots**
![Clip](https://user-images.githubusercontent.com/24726299/92644912-b7518c80-f2ec-11ea-8a80-083d07be7a72.png)

**Environment:**
 - Vim version [8.2.1632](https://github.com/vim/vim-win32-installer/releases/download/v8.2.1632/gvim_8.2.1632_x86_signed.zip)
 - OS: Windows Home 1803
 - Terminal: GUI

**Comment by lkintact** [comment]
There seems to be a side issue that comes with this one:

If I don't include `--cmd "set encoding=utf-8"` at step 2, i.e. run 

`gvim --clean --cmd "language messages English" "[id-01] file.txt"`,

the message text for E944 at step 4 is not in English despite the `--cmd "language messages English"` switch:

![Clip](https://user-images.githubusercontent.com/24726299/92649377-99d3f100-f2f3-11ea-87d2-c779f2e09097.png)

(That's Russian, I run a Russian version of the OS). Notice that the "Press ENTER ..." prompt is in English, in compliance with the switch value.

**Comment by rkta** [comment]
VIM - Vi IMproved 8.2 (2019 Dec 12, compiled Jan 30 2021 11:42:28)
Included patches: 1-2427


Some more details why this happens:

'%' is replaced in 'eval_vars()' with 'curbuf->b_name'. In 'b_name'
special characters are not escaped.

Later the command is passed to 'gen_expand_wildcards()', which expects
all special chars to be escaped. Here '[id-1]' is correctly recognized
as a wildcard. Which finally leads to the error message.


Possible solution would be to not replace '%' with the literal file
name, but with a quoted or escaped one. I'm not sure how to do this
platform independent.


Here is backtrace from when the error is returned. In this case
'eval_vars()' was called before entering frame `#14` .

```
Breakpoint 1, nfa_regatom () at regexp_nfa.c:1839
1839                                EMSG_RET_FAIL(_(e_reverse_range));
(gdb) bt
#0  nfa_regatom () at regexp_nfa.c:1839
#1  0x000000000056f461 in nfa_regpiece () at regexp_nfa.c:2027
#2  0x000000000056fcb9 in nfa_regconcat () at regexp_nfa.c:2271
#3  0x000000000056fd45 in nfa_regbranch () at regexp_nfa.c:2304
#4  0x0000000000570009 in nfa_reg (paren=0x0) at regexp_nfa.c:2365
#5  0x0000000000570260 in re2post () at regexp_nfa.c:2784
#6  0x00000000005780b8 in nfa_regcomp (expr=0x9e0ba0 "^[id-1]$", re_flags=0x9) at regexp_nfa.c:7294
#7  0x00000000005784cf in vim_regcomp (expr_arg=0x9e0ba0 "^[id-1]$", re_flags=0x1) at regexp.c:2660
#8  0x00000000004a38c9 in unix_expandpath (gap=0x7fffffffd240, path=0x9e0fef "/tmp/vim/[id-1]", wildoff=0x0, flags=0x6, didstar=0x0) at filepath.c:3569
#9  0x0000000000537471 in mch_expandpath (gap=0x7fffffffd240, path=0x9e0fef "/tmp/vim/[id-1]", flags=0x6) at os_unix.c:6465
#10 0x00000000004a4198 in gen_expand_wildcards (num_pat=0x1, pat=0xa1a4f0, num_file=0x7fffffffd3b0, file=0x7fffffffd3a8, flags=0x6) at filepath.c:3864
#11 0x00000000004a2fda in expand_wildcards (num_pat=0x1, pat=0xa1a4f0, num_files=0x7fffffffd3b0, files=0x7fffffffd3a8, flags=0x6) at filepath.c:3018
#12 0x000000000040584a in get_arglist_exp (str=0x9e0fef "/tmp/vim/[id-1]", fcountp=0x7fffffffd3b0, fnamesp=0x7fffffffd3a8, wig=0x1) at arglist.c:317
#13 0x0000000000550b78 in vgr_process_args (eap=0x7fffffffd680, args=0x7fffffffd390) at quickfix.c:6057
#14 0x00000000005511d9 in ex_vimgrep (eap=0x7fffffffd680) at quickfix.c:6275
#15 0x000000000047f9ba in do_one_cmd (cmdlinep=0x7fffffffdea0, flags=0x0, cstack=0x7fffffffd870, fgetline=0x49342e <getexline>, cookie=0x0) at ex_docmd.c:2585
#16 0x000000000047cdb4 in do_cmdline (cmdline=0x0, fgetline=0x49342e <getexline>, cookie=0x0, flags=0x0) at ex_docmd.c:1000
#17 0x00000000005101d0 in nv_colon (cap=0x7fffffffdfd0) at normal.c:3383
#18 0x000000000050c5a9 in normal_cmd (oap=0x7fffffffe0b0, toplevel=0x1) at normal.c:1100
#19 0x000000000067162d in main_loop (cmdwin=0x0, noexmode=0x0) at main.c:1475
#20 0x0000000000670ade in vim_main2 () at main.c:865
#21 0x0000000000670336 in main (argc=0x2, argv=0x7fffffffe2b8) at main.c:412
```

**Comment by brammool** [comment]
Fixed with patch 8.2.2430
